File: python/tests/test_hypotheses_joint.py
import polars as pl
import pytest
import statsmodels.formula.api as smf
from polars.testing import assert_frame_equal

from marginaleffects import *
from tests.helpers import mtcars


mod = smf.ols("am ~ hp + wt + disp", data=mtcars).fit()

mod_without_intercept = smf.ols("am ~ 0 + hp + wt + disp", data=mtcars).fit()


@pytest.mark.parametrize(
    "kwargs, expected_file",
    [
        ({"joint": ["hp", "wt"]}, "test_hypotheses_joint_01.csv"),
        (
            {"joint": ["hp", "disp"], "joint_test": "chisq"},
            "test_hypotheses_joint_02.csv",
        ),
        ({"joint": [1, 2]}, "test_hypotheses_joint_03.csv"),
        ({"joint": [0, 1, 2], "hypothesis": [1, 2, 3]}, "test_hypotheses_joint_04.csv"),
        (
            {"joint": ["Intercept", "disp", "wt"], "hypothesis": 4},
            "test_hypotheses_joint_05.csv",
        ),
    ],
)
def test_hypotheses_joint(kwargs, expected_file):
    hypo_py = hypotheses(mod, **kwargs)
    hypo_r = pl.read_csv(f"tests/r/{expected_file}").rename({"p.value": "p_value"})
    hypo_r = hypo_r.cast({"p_value": pl.Float64})
    assert_frame_equal(hypo_py.to_polars(), hypo_r)


@pytest.mark.parametrize(
    "kwargs, expected_file",
    [
        ({"joint": [0, 1, 2]}, "test_hypotheses_joint_06.csv"),
        ({"joint": ["hp", "wt"]}, "test_hypotheses_joint_07.csv"),
    ],
)
def test_hypotheses_joint_2(kwargs, expected_file):
    hypo_py = hypotheses(mod_without_intercept, **kwargs)
    hypo_r = pl.read_csv(f"tests/r/{expected_file}").rename({"p.value": "p_value"})
    assert_frame_equal(hypo_py.to_polars(), hypo_r)


@pytest.mark.parametrize(
    "kwargs, expected_file",
    [
        ({"joint": "hp|wt"}, "test_hypotheses_joint_01.csv"),
        ({"joint": "^..$"}, "test_hypotheses_joint_01.csv"),
        (
            {"joint": "p$", "joint_test": "chisq"},
            "test_hypotheses_joint_02.csv",
        ),  # matches hp and disp
    ],
)
def test_hypotheses_joint_regex(kwargs, expected_file):
    hypo_py = hypotheses(mod, **kwargs)
    hypo_r = pl.read_csv(f"tests/r/{expected_file}").rename({"p.value": "p_value"})
    assert_frame_equal(hypo_py.to_polars(), hypo_r)

File: python/tests/r/test_hypotheses_joint.R
source(here::here("tests/r/load.R"))

# Read the same vendored CSV as tests/test_hypotheses_joint.py.
mtcars_dat <- fread(here("tests/data/mtcars.csv"), na.strings = c("NA", ""))

mod <- lm(am ~ hp + wt + disp, data = mtcars_dat)
mod_without_intercept <- lm(am ~ 0 + hp + wt + disp, data = mtcars_dat)

# `joint` indices are 0-based in Python and 1-based in R, so the R positions
# below are the Python positions in tests/test_hypotheses_joint.py plus one.
save <- function(x, file) {
    write.csv(x, here(file.path("tests/r", file)), row.names = FALSE)
}

save(hypotheses(mod, joint = c("hp", "wt")), "test_hypotheses_joint_01.csv")
save(
    hypotheses(mod, joint = c("hp", "disp"), joint_test = "chisq"),
    "test_hypotheses_joint_02.csv"
)
save(hypotheses(mod, joint = c(2, 3)), "test_hypotheses_joint_03.csv")
save(
    hypotheses(mod, joint = c(1, 2, 3), hypothesis = c(1, 2, 3)),
    "test_hypotheses_joint_04.csv"
)
save(
    hypotheses(mod, joint = c("(Intercept)", "disp", "wt"), hypothesis = 4),
    "test_hypotheses_joint_05.csv"
)
save(
    hypotheses(mod_without_intercept, joint = c(1, 2, 3)),
    "test_hypotheses_joint_06.csv"
)
save(
    hypotheses(mod_without_intercept, joint = c("hp", "wt")),
    "test_hypotheses_joint_07.csv"
)
